cachi2/core/package_managers/pip.py

```diff
@@ -1410,12 +1423,26 @@ def _download_dependencies(
 
         if require_hashes or req.kind == "url":
             hashes = req.hashes or [req.qualifiers["cachito_hash"]]
-            _verify_hash(download_info["path"], hashes)
+            _verify_hash(download_info["path"], list(map(_to_checksum_info, hashes)))
 
         download_info["kind"] = req.kind
-        downloads.append(download_info)
+        downloaded.append(download_info)
+
+    if allow_binary:
+        log.info("Downloading %d wheel(s) ...", len(to_download))
+        files: dict[str, Union[str, PathLike[str]]] = {pkg.url: pkg.path for pkg in to_download}
+        asyncio.run(async_download_files(files, get_config().concurrency_limit))
+
+        for pkg in to_download:
+            try:
+                _verify_hash(pkg.path, pkg.checksums)
+            except PackageRejected:
+                pkg.path.unlink()
+                log.warning("Computed hash does not match any of the provided hashes")
```

Comment:
What does the cachi2 log output look like when this happens? Will the user understand what's going on and why?

Reply:
It would be helpful to see the actual output if you could post an example here

Reply:
```
2023-08-14 12:49:42,708 INFO Downloading 1 wheel(s) ...
2023-08-14 12:49:42,809 INFO Verifying checksum of aiowsgi-0.7-py3-none-any.whl
2023-08-14 12:49:42,809 WARNING aiowsgi-0.7-py3-none-any.whl: sha256 checksum does not match (got: caa338f95d6eb0908015b63cf2147ced24a8f3c88b6254f79a3cbd814b1fdcc5)
2023-08-14 12:49:42,810 WARNING The file is removed from the output directory
2023-08-14 12:49:42,810 INFO Any existing wheels were not re-downloaded
2023-08-14 12:49:42,828 INFO All dependencies fetched successfully \o/
```

Reply:
the first warning is from `_log_mismatches` function
the second I added by myself

Reply:
Nice, that looks fairly reasonable. But it would be nice to differentiate between verifying the PyPI-reported hashes vs. the user-specified hashes (related to https://github.com/containerbuildsystem/cachi2/pull/255#discussion_r1294452974)